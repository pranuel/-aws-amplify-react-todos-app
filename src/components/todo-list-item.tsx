import React, { ChangeEvent } from "react";
import { isEnterKey } from "../helpers/keyboard.helper";
import { Todo } from "../todos/todo.model";

export interface TodoListItemProps {
  todo: Todo;
  editTodo?: Todo;
  onEditTodo: (todo: Todo) => void;
  onTodoIsDoneToggled: (todo: Todo, isDone: boolean) => void;
  onSaveTodo: (todo: Todo) => void;
  onDeleteTodo: (todo: Todo) => void;
  onEditingTodoDescriptionChanged: (
    editTodo: Todo,
    description: string,
  ) => void;
}

export class TodoListItem extends React.Component<TodoListItemProps, {}> {
  public render() {
    const { todo, editTodo } = this.props;
    return (
      <li
        className={this.getTodoListItemClassName(todo, editTodo)}
        onDoubleClick={this.onDoubleClick}
      >
        <div className="view">
          <input
            className="toggle"
            type="checkbox"
            data-testid="is-done-toggle"
            checked={todo.isDone}
            onChange={this.onTodoIsDoneToggled}
          />
          <label>{todo.description}</label>
          <button
            className="destroy"
            data-testid="delete-button"
            onClick={this.onDeleteTodoClicked}
          />
        </div>
        <input
          className="edit"
          value={!!editTodo ? editTodo.description : ""}
          data-testid="edit-input"
          onKeyDown={this.onEditInputKeyDown}
          onBlur={this.onEditInputBlur}
          onChange={this.onEditInputChanged}
        />
      </li>
    );
  }

  private onEditInputChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (!this.props.editTodo) {
      return;
    }
    this.props.onEditingTodoDescriptionChanged(
      this.props.editTodo,
      event.target.value,
    );
  }

  private onDeleteTodoClicked = () => {
    this.props.onDeleteTodo(this.props.todo);
  }

  private onEditInputBlur = () => {
    if (!this.props.editTodo) {
      return;
    }
    this.props.onSaveTodo(this.props.editTodo);
  }

  private onTodoIsDoneToggled = (event: ChangeEvent<HTMLInputElement>) => {
    this.props.onTodoIsDoneToggled(this.props.todo, event.target.checked);
  }

  private onEditInputKeyDown = (
    event: React.KeyboardEvent<HTMLInputElement>,
  ) => {
    if (!isEnterKey(event.key) || !this.props.editTodo) {
      return;
    }
    this.props.onSaveTodo(this.props.editTodo);
  }

  private onDoubleClick = () => {
    this.props.onEditTodo(this.props.todo);
  }

  private getTodoListItemClassName(todo: Todo, editTodo?: Todo): string {
    const classNames = [];
    if (todo.isDone) {
      classNames.push("completed");
    }
    if (!!editTodo && editTodo.id === todo.id) {
      classNames.push("editing");
    }
    return classNames.join(" ");
  }
}
