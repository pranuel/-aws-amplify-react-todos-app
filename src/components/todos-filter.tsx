import React from "react";
import { TodoFilter } from "../todos/todo-filter.model";

interface TodosFilterProps {
  todoFilter: TodoFilter;
  onShowAllTodosClicked: () => void;
  onShowActiveTodosClicked: () => void;
  onShowCompletedTodosClicked: () => void;
}

export class TodosFilter extends React.Component<TodosFilterProps, {}> {
  public render() {
    return (
      <ul className="filters">
        <li>
          <a
            href="#/"
            onClick={this.props.onShowAllTodosClicked}
            className={this.getTodosFilterClassName(TodoFilter.All)}
          >
            All
          </a>
        </li>
        <span />
        <li>
          <a
            href="#/"
            onClick={this.props.onShowActiveTodosClicked}
            className={this.getTodosFilterClassName(TodoFilter.Active)}
          >
            Active
          </a>
        </li>
        <span />
        <li>
          <a
            href="#/"
            onClick={this.props.onShowCompletedTodosClicked}
            className={this.getTodosFilterClassName(TodoFilter.Completed)}
          >
            Completed
          </a>
        </li>
      </ul>
    );
  }

  private getTodosFilterClassName = (elementTodoFilter: TodoFilter) => {
    return this.props.todoFilter === elementTodoFilter ? "selected" : undefined;
  }
}
