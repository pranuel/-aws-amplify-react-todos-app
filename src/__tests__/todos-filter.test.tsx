import { fireEvent, render } from "@testing-library/react";
import React from "react";
import renderer from "react-test-renderer";
import { TodosFilter } from "../components/todos-filter";
import { TodoFilter } from "../todos/todo-filter.model";

describe("TodosFilter", () => {
  it("should highlight the selected filter item", () => {
    const props = {
      currentTodoFilter: TodoFilter.Active,
      onShowAllTodosClicked: jest.fn(),
      onShowActiveTodosClicked: jest.fn(),
      onShowCompletedTodosClicked: jest.fn(),
    };
    const tree = renderer.create(<TodosFilter {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should call the provided callback if a filter item is clicked", () => {
    // arrange
    const props = {
      currentTodoFilter: TodoFilter.Active,
      onShowAllTodosClicked: jest.fn(),
      onShowActiveTodosClicked: jest.fn(),
      onShowCompletedTodosClicked: jest.fn(),
    };
    const onShowCompletedTodosClickedSpy = spyOn(
      props,
      "onShowCompletedTodosClicked",
    );
    const { getByText } = render(<TodosFilter {...props} />);
    // act
    fireEvent.click(getByText("Completed"));
    // assert
    expect(onShowCompletedTodosClickedSpy).toHaveBeenCalled();
  });
});
