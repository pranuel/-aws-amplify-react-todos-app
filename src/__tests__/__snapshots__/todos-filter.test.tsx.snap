// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosFilter should highlight the selected filter item 1`] = `
<ul
  className="filters"
>
  <li>
    <a
      href="#/"
      onClick={[MockFunction]}
    >
      All
    </a>
  </li>
  <span />
  <li>
    <a
      className="selected"
      href="#/"
      onClick={[MockFunction]}
    >
      Active
    </a>
  </li>
  <span />
  <li>
    <a
      href="#/"
      onClick={[MockFunction]}
    >
      Completed
    </a>
  </li>
</ul>
`;
